File: avenger-vega/src/marks/symbol.rs
use crate::error::AvengerVegaError;
use crate::marks::mark::{VegaMarkContainer, VegaMarkItem};
use crate::marks::values::{CssColorOrGradient, StrokeDashSpec};
use avenger_common::types::{ColorOrGradient, Gradient, StrokeCap, StrokeJoin, SymbolShape};
use avenger_common::value::ScalarOrArray;
use avenger_scenegraph::marks::group::{Clip, SceneGroup};
use avenger_scenegraph::marks::line::SceneLineMark;
use avenger_scenegraph::marks::mark::SceneMark;
use avenger_scenegraph::marks::symbol::SceneSymbolMark;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use lyon_extra::parser::ParseError;

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VegaSymbolItem {
    pub x: Option<f32>,
    pub y: Option<f32>,
    pub fill: Option<CssColorOrGradient>,
    pub opacity: Option<f32>,
    pub fill_opacity: Option<f32>,
    pub size: Option<f32>,
    pub shape: Option<String>,
    pub stroke: Option<CssColorOrGradient>,
    pub stroke_width: Option<f32>,
    pub stroke_cap: Option<StrokeCap>,
    pub stroke_join: Option<StrokeJoin>,
    pub stroke_dash: Option<StrokeDashSpec>,
    pub stroke_opacity: Option<f32>,
    pub angle: Option<f32>,
    pub zindex: Option<i32>,
}

impl VegaMarkItem for VegaSymbolItem {}

impl VegaMarkContainer<VegaSymbolItem> {
    pub fn to_scene_graph(&self, force_clip: bool) -> Result<SceneMark, AvengerVegaError> {
        // Get shape of first item and use that for all items for now
        let first = self.items.first();
        let first_shape = first
            .and_then(|item| item.shape.clone())
            .unwrap_or("circle".to_string());

        // Handle special case of sybols with shape == "stroke". This happens when lines are
        // used in legends. We convert these to a group of regular line marks
        if first_shape == "stroke" {
            // stroke symbols are converted to a group of lines
            let mut line_marks: Vec<SceneMark> = Vec::new();
            for item in &self.items {
                let mut gradients = Vec::<Gradient>::new();
                let width = item.size.unwrap_or(100.0).sqrt();
                let stroke = if let Some(c) = &item.stroke {
                    let base_opacity = item.opacity.unwrap_or(1.0);
                    let stroke_opacity = item.stroke_opacity.unwrap_or(1.0) * base_opacity;
                    c.to_color_or_grad(stroke_opacity, &mut gradients)?
                } else {
                    ColorOrGradient::Color([0.0, 0.0, 0.0, 0.0])
                };
                let x = item.x.unwrap_or(0.0);
                let y = item.y.unwrap_or(0.0);
                let mark = SceneLineMark {
                    name: "".to_string(),
                    clip: false,
                    zindex: self.zindex,
                    len: 2,
                    x: ScalarOrArray::new_array(vec![x - width / 2.0, x + width / 2.0]),
                    y: ScalarOrArray::new_scalar(y),
                    stroke,
                    stroke_width: item.stroke_width.unwrap_or(1.0),
                    stroke_cap: item.stroke_cap.unwrap_or_default(),
                    stroke_join: item.stroke_join.unwrap_or_default(),
                    stroke_dash: item
                        .stroke_dash
                        .clone()
                        .map(|d| Ok::<Vec<f32>, AvengerVegaError>(d.to_array()?.to_vec()))
                        .transpose()?,
                    gradients,
                    ..Default::default()
                };
                line_marks.push(SceneMark::Line(mark));
            }
            return Ok(SceneMark::Group(SceneGroup {
                name: "symbol_line_legend".to_string(),
                origin: [0.0, 0.0],
                clip: Clip::None,
                marks: line_marks,
                gradients: vec![],
                fill: None,
                stroke: None,
                stroke_width: None,
                stroke_offset: None,
                zindex: None,
            }));
        }

        let first_has_stroke = first.map(|item| item.stroke.is_some()).unwrap_or(false);

        // Only include stroke_width if there is a stroke color
        let stroke_width = if first_has_stroke {
            first.and_then(|item| item.stroke_width)
        } else {
            None
        };

        // Init mark with scalar defaults
        let mut mark = SceneSymbolMark {
            stroke_width,
            clip: self.clip || force_clip,
            zindex: self.zindex,
            ..Default::default()
        };

        if let Some(name) = &self.name {
            mark.name = name.clone();
        }

        // Override values with vectors
        let len = self.items.len();
        mark.len = len as u32;

        // Init vector for each encoding channel
        let mut x = Vec::<f32>::with_capacity(len);
        let mut y = Vec::<f32>::with_capacity(len);
        let mut fill = Vec::<ColorOrGradient>::with_capacity(len);
        let mut size = Vec::<f32>::with_capacity(len);
        let mut stroke = Vec::<ColorOrGradient>::with_capacity(len);
        let mut angle = Vec::<f32>::with_capacity(len);
        let mut zindex = Vec::<i32>::with_capacity(len);
        let mut gradients = Vec::<Gradient>::with_capacity(len);

        let mut shape_strings = Vec::<String>::with_capacity(len);
        let mut shape_index = Vec::<usize>::with_capacity(len);

        // For each item, append explicit values to corresponding vector
        for item in &self.items {
            x.push(item.x.unwrap_or(0.0));
            y.push(item.y.unwrap_or(0.0));

            let base_opacity = item.opacity.unwrap_or(1.0);
            if let Some(v) = &item.fill {
                let fill_opacity = item.fill_opacity.unwrap_or(1.0) * base_opacity;
                fill.push(v.to_color_or_grad(fill_opacity, &mut gradients)?);
            }

            if let Some(s) = item.size {
                size.push(s);
            }

            if let Some(v) = &item.stroke {
                let stroke_opacity = item.stroke_opacity.unwrap_or(1.0) * base_opacity;
                stroke.push(v.to_color_or_grad(stroke_opacity, &mut gradients)?);
            }
            if let Some(v) = item.angle {
                angle.push(v);
            }
            if let Some(v) = item.zindex {
                zindex.push(v);
            }
            if let Some(shape_string) = &item.shape {
                if let Some(pos) = shape_strings.iter().position(|s| s == shape_string) {
                    // Already have shape
                    shape_index.push(pos);
                } else {
                    // Add shape
                    let pos = shape_strings.len();
                    shape_strings.push(shape_string.clone());
                    shape_index.push(pos);
                }
            }
        }

        if x.len() == len {
            mark.x = ScalarOrArray::new_array(x);
        }
        if y.len() == len {
            mark.y = ScalarOrArray::new_array(y);
        }
        if fill.len() == len {
            mark.fill = ScalarOrArray::new_array(fill);
        }
        if size.len() == len {
            mark.size = ScalarOrArray::new_array(size);
        }
        if stroke.len() == len {
            mark.stroke = ScalarOrArray::new_array(stroke);
        }
        if angle.len() == len {
            mark.angle = ScalarOrArray::new_array(angle);
        }
        if zindex.len() == len {
            let mut indices: Vec<usize> = (0..len).collect();
            indices.sort_by_key(|i| zindex[*i]);
            mark.indices = Some(Arc::new(indices));
        }
        if shape_index.len() == len {
            mark.shape_index = ScalarOrArray::new_array(shape_index);
            mark.shapes = shape_strings
                .iter()
                .map(|s| SymbolShape::from_vega_str(s))
                .collect::<Result<Vec<SymbolShape>, ParseError>>()?;
        }

        // Add gradients
        mark.gradients = gradients;

        Ok(SceneMark::Symbol(mark))
    }
}

File: avenger-vega-renderer/js/marks/symbol.js
import {SymbolMark, GroupMark, LineMark} from "../../pkg/avenger_wasm.js";
import {encodeSimpleArray} from "./util.js";


/**
 * @typedef {Object} SymbolItem
 * @property {number} strokeWidth
 * @property {"butt"|"round"|"square"} strokeCap
 * @property {"bevel"|"miter"|"round"} strokeJoin
 * @property {string|number[]} strokeDash
 * @property {string|object} fill
 * @property {string|object} stroke
 * @property {number} x
 * @property {number} y
 * @property {number} size
 * @property {number} opacity
 * @property {number} strokeOpacity
 * @property {number} fillOpacity
 * @property {string} shape
 * @property {number} angle
 * @property {number} zindex
 */

/**
 * @typedef {Object} SymbolMarkSpec
 * @property {"symbol"} marktype
 * @property {boolean} clip
 * @property {boolean} interactive
 * @property {SymbolItem[]} items
 * @property {string} name
 * @property {string} role
 * @property {number} zindex
 */

/**
 * @param {SymbolMarkSpec} vegaSymbolMark
 * @param {boolean} force_clip
 * @returns {SymbolMark}
 */
export function importSymbol(vegaSymbolMark, force_clip) {
    const items = vegaSymbolMark.items;
    const len = items.length;

    const symbolMark = new SymbolMark(
        len, vegaSymbolMark.clip || force_clip, vegaSymbolMark.name, vegaSymbolMark.zindex
    );

    // Handle empty mark
    if (len === 0) {
        return symbolMark;
    }

    const firstItem = items[0];
    // Only include stroke_width if there is a stroke color
    const firstHasStroke = firstItem.stroke != null;
    let strokeWidth;
    if (firstHasStroke) {
        strokeWidth = firstItem.strokeWidth ?? 1;
    }
    symbolMark.set_stroke_width(strokeWidth);

    // Semi-required values get initialized
    const x = new Float32Array(len).fill(0);
    const y = new Float32Array(len).fill(0);

    const fill = new Array(len).fill("");;
    let anyFill = false;
    let anyFillIsGradient = false;

    const size = new Float32Array(len).fill(20);
    let anySize = false;

    const stroke = new Array(len).fill("");;
    let anyStroke = false;
    let anyStrokeIsGradient = false;

    const angle = new Float32Array(len).fill(0);
    let anyAngle = false;

    const zindex = new Int32Array(len).fill(0);
    let anyZindex = false;

    const fillOpacity = new Float32Array(len).fill(1);
    const strokeOpacity = new Float32Array(len).fill(1);

    const shapes = new Array(len);
    let anyShape = false;

    items.forEach((item, i) => {
        x[i] = item.x ?? 0;
        y[i] = item.y ?? 0;

        const baseOpacity = item.opacity ?? 1;
        fillOpacity[i] = (item.fillOpacity ?? 1) * baseOpacity;
        strokeOpacity[i] = (item.strokeOpacity ?? 1) * baseOpacity;

        if (item.fill != null) {
            fill[i] = item.fill;
            anyFill ||= true;
            anyFillIsGradient ||= typeof item.fill === "object";
        }

        if (item.size != null) {
            size[i] = item.size;
            anySize = true;
        }

        if (item.stroke != null) {
            stroke[i] = item.stroke;
            anyStroke ||= true;
            anyStrokeIsGradient ||= typeof item.stroke === "object";
        }

        if (item.angle != null) {
            angle[i] = item.angle;
            anyAngle ||= true;
        }

        if (item.zindex != null) {
            zindex[i] = item.zindex;
            anyZindex ||= true;
        }

        if (item.shape != null) {
            shapes[i] = item.shape;
            anyShape ||= true;
        }
    })

    symbolMark.set_xy(x, y);

    if (anyFill) {
        if (anyFillIsGradient) {
            symbolMark.set_fill_gradient(fill, fillOpacity);
        } else {
            const encoded = encodeSimpleArray(fill);
            symbolMark.set_fill(encoded.values, encoded.indices, fillOpacity);
        }
    }

    if (anySize) {
        symbolMark.set_size(size);
    }

    if (anyStroke) {
        if (anyStrokeIsGradient) {
            symbolMark.set_stroke_gradient(stroke, strokeOpacity);
        } else {
            const encoded = encodeSimpleArray(stroke);
            symbolMark.set_stroke(encoded.values, encoded.indices, strokeOpacity);
        }
    }

    if (anyAngle) {
        symbolMark.set_angle(angle);
    }

    if (anyZindex) {
        symbolMark.set_zindex(zindex);
    }

    if (anyShape) {
        const encoded = encodeSimpleArray(shapes);
        symbolMark.set_shape(encoded.values, encoded.indices);
    }

    return symbolMark;
}


/**
 * Handle special case of symbols with shape == "stroke". This happens when lines are
 * sed in legends. We convert these to a group of regular line marks
 * @param {SymbolMarkSpec} vegaSymbolMark
 * @param {boolean} force_clip
 * @returns {GroupMark}
 */
export function importStrokeLegend(vegaSymbolMark, force_clip) {
    const groupMark = new GroupMark(
        0, 0, "symbol_line_legend", undefined, undefined
    );

    for (let item of vegaSymbolMark.items) {
        let width = Math.sqrt(item.size ?? 100.0);
        let x = item.x ?? 0;
        let y = item.y ?? 0;
        const lineMark = new LineMark(
            2, vegaSymbolMark.clip || force_clip, undefined, vegaSymbolMark.zindex
        );

        lineMark.set_xy(
            new Float32Array([x - width / 2.0, x + width / 2.0]),
            new Float32Array([y, y])
        )

        lineMark.set_stroke(item.stroke ?? "", (item.opacity ?? 1) * (item.strokeOpacity ?? 1));
        if (item.strokeWidth != null) {
            lineMark.set_stroke_width(item.strokeWidth);
        }

        if (item.strokeCap != null) {
            lineMark.set_stroke_cap(item.strokeCap);
        }

        if (item.strokeJoin != null) {
            lineMark.set_stroke_join(item.strokeJoin);
        }

        if (item.strokeDash != null) {
            lineMark.set_stroke_dash(item.strokeDash);
        }

        groupMark.add_line_mark(lineMark);
    }

    return groupMark
}
